False positive with public protocol

With this code:

```swift
public protocol Foo {
    var bar: Bool { get }
}

func bind<T>(_ type: T.Type, to instantiator: @escaping @autoclosure () -> T) -> () -> T {
    fatalError("just here for the type")
}

struct Bar: Foo {
    let bar = false
}

/// Provides access to wheelchair user preference.
public let getFoo = bind(Foo.self, to: Bar())
```

You get this false positive:

```
protocol/Sources/protocol/protocol.swift:1:17: warning: Protocol 'Foo' is declared public, but not used outside of protocol
```

But if you remove that you get this error:

```
protocol/Sources/protocol/protocol.swift:14:12: error: constant cannot be declared public because its type '() -> any Foo' uses an internal type
public let getFoo = bind(Foo.self, to: Bar())
           ^
```

example project: [protocol.zip](https://github.com/peripheryapp/periphery/files/10438618/protocol.zip)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@
 
 - `COMPILER_INDEX_STORE_ENABLE` is now forcefully enabled as it's required for indexing.
 - Fix false positive where a `typealias` is extended but otherwise unused.
+- Fix redundant accessibility analysis for function metatype arguments.
 
 ## 2.12.1 (2023-03-04)
 

--- a/Sources/PeripheryKit/Indexer/Declaration.swift
+++ b/Sources/PeripheryKit/Indexer/Declaration.swift
@@ -177,6 +177,7 @@ public final class Declaration {
     public var declaredType: String?
     public var letShorthandIdentifiers: Set<String> = []
     public var hasCapitalSelfFunctionCall: Bool = false
+    public var hasGenericFunctionReturnedMetatypeParameters: Bool = false
 
     public var parent: Declaration?
     var related: Set<Reference> = []

--- a/Sources/PeripheryKit/Indexer/Reference.swift
+++ b/Sources/PeripheryKit/Indexer/Reference.swift
@@ -7,18 +7,17 @@ public final class Reference {
         case genericRequirementType
         case inheritedClassType
         case refinedProtocolType
+        case variableInitFunctionCall
+        case functionCallMetatypeArgument
         case unknown
 
-        static var publiclyExposableRoles: [Role] {
-            return [
-                .varType,
-                .returnType,
-                .parameterType,
-                .genericParameterType,
-                .genericRequirementType,
-                .inheritedClassType,
-                .refinedProtocolType
-            ]
+        var isPubliclyExposable: Bool {
+            switch self {
+            case .varType, .returnType, .parameterType, .genericParameterType, .genericRequirementType, .inheritedClassType, .refinedProtocolType, .functionCallMetatypeArgument:
+                return true
+            default:
+                return false
+            }
         }
     }
 

--- a/Sources/PeripheryKit/Indexer/SwiftIndexer.swift
+++ b/Sources/PeripheryKit/Indexer/SwiftIndexer.swift
@@ -372,6 +372,7 @@ public final class SwiftIndexer: Indexer {
                 decl.declaredType = result.variableType
                 decl.letShorthandIdentifiers = result.letShorthandIdentifiers
                 decl.hasCapitalSelfFunctionCall = result.hasCapitalSelfFunctionCall
+                decl.hasGenericFunctionReturnedMetatypeParameters = result.hasGenericFunctionReturnedMetatypeParameters
 
                 for ref in decl.references.union(decl.related) {
                     if result.inheritedTypeLocations.contains(ref.location) {
@@ -390,6 +391,10 @@ public final class SwiftIndexer: Indexer {
                         ref.role = .genericParameterType
                     } else if result.genericConformanceRequirementLocations.contains(ref.location) {
                         ref.role = .genericRequirementType
+                    } else if result.variableInitFunctionCallLocations.contains(ref.location) {
+                        ref.role = .variableInitFunctionCall
+                    } else if result.functionCallMetatypeArgumentLocations.contains(ref.location) {
+                        ref.role = .functionCallMetatypeArgument
                     }
                 }
             }

--- a/Sources/PeripheryKit/SourceGraph/Mutators/RedundantExplicitPublicAccessibilityMarker.swift
+++ b/Sources/PeripheryKit/SourceGraph/Mutators/RedundantExplicitPublicAccessibilityMarker.swift
@@ -74,8 +74,30 @@ final class RedundantExplicitPublicAccessibilityMarker: SourceGraphMutator {
 
     private func isExposedPubliclyByAnotherDeclaration(_ decl: Declaration) -> Bool {
         let referenceDecls = graph.references(to: decl)
-            .filter { Reference.Role.publiclyExposableRoles.contains($0.role) }
-            .compactMap { $0.parent }
+            .compactMap {
+                if $0.role.isPubliclyExposable {
+                    if case .functionCallMetatypeArgument = $0.role {
+                        // This reference is a function metatype argument. If the called function
+                        // also has generic metatype parameters, and returns any of the generic
+                        // types, then we must assume the argument type is publicly accessible.
+                        // Note that this isn't 100% accurate, as the metatype argument may not
+                        // correspond to a generic type. However, it's too complex to match up
+                        // the call site arguments with function parameters, as parameters with
+                        // default types can cause misalignment between the positions of the two.
+                        if let functionRef = $0.parent?.references.first(where: { $0.role == .variableInitFunctionCall }),
+                           let functionDecl = graph.explicitDeclaration(withUsr: functionRef.usr),
+                           functionDecl.hasGenericFunctionReturnedMetatypeParameters {
+                            return $0.parent
+                        }
+
+                        return nil
+                    }
+
+                    return $0.parent
+                }
+
+                return nil
+            }
 
         return referenceDecls.contains { refDecl in
             refDecl.accessibility.value == .public || refDecl.accessibility.value == .open

--- a/Sources/PeripheryKit/Syntax/DeclarationSyntaxVisitor.swift
+++ b/Sources/PeripheryKit/Syntax/DeclarationSyntaxVisitor.swift
@@ -15,8 +15,11 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
         inheritedTypeLocations: Set<SourceLocation>,
         genericParameterLocations: Set<SourceLocation>,
         genericConformanceRequirementLocations: Set<SourceLocation>,
+        variableInitFunctionCallLocations: Set<SourceLocation>,
+        functionCallMetatypeArgumentLocations: Set<SourceLocation>,
         letShorthandIdentifiers: Set<String>,
-        hasCapitalSelfFunctionCall: Bool
+        hasCapitalSelfFunctionCall: Bool,
+        hasGenericFunctionReturnedMetatypeParameters: Bool
     )
 
     var letShorthandWorkaroundEnabled: Bool = false
@@ -200,6 +203,7 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
             if binding.pattern.is(IdentifierPatternSyntax.self) {
                 let closureSignature = binding.initializer?.value.as(ClosureExprSyntax.self)?.signature
                 let closureParameters = closureSignature?.input?.as(ParameterClauseSyntax.self)
+                let functionCallExpr = binding.initializer?.value.as(FunctionCallExprSyntax.self)
                 parse(
                     modifiers: node.modifiers,
                     attributes: node.attributes,
@@ -207,35 +211,15 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
                     variableType: binding.typeAnnotation?.type,
                     parameterClause: closureParameters,
                     returnClause: closureSignature?.output,
+                    variableInitFunctionCallExpr: functionCallExpr,
                     at: binding.positionAfterSkippingLeadingTrivia
                 )
             } else if let tuplePatternSyntax = binding.pattern.as(TuplePatternSyntax.self) {
-                // Destructuring binding.
-                let positions = tuplePatternSyntax.elements.map { $0.positionAfterSkippingLeadingTrivia }
-
-                if let typeSyntax = binding.typeAnnotation?.type {
-                    if let tupleType = typeSyntax.as(TupleTypeSyntax.self) {
-                        // Inspect elements individually, and associate each type with its corresponding identifier.
-                        for (position, elem) in zip(positions, tupleType.elements) {
-                            parse(
-                                modifiers: node.modifiers,
-                                attributes: node.attributes,
-                                trivia: node.leadingTrivia,
-                                variableType: elem.type,
-                                at: position
-                            )
-                        }
-                    }
-                } else {
-                    for position in positions {
-                        parse(
-                            modifiers: node.modifiers,
-                            attributes: node.attributes,
-                            trivia: node.leadingTrivia,
-                            at: position
-                        )
-                    }
-                }
+                visitVariableTupleBinding(
+                    node: node,
+                    pattern: tuplePatternSyntax,
+                    typeTuple: binding.typeAnnotation?.type.as(TupleTypeSyntax.self)?.elements,
+                    initializerTuple: binding.initializer?.value.as(TupleExprSyntax.self)?.elementList)
             } else {
                 parse(
                     modifiers: node.modifiers,
@@ -243,6 +227,33 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
                     trivia: node.leadingTrivia,
                     at: binding.positionAfterSkippingLeadingTrivia
                 )
+            }
+        }
+    }
+
+    func visitVariableTupleBinding(node: VariableDeclSyntax, pattern: TuplePatternSyntax, typeTuple: TupleTypeElementListSyntax?, initializerTuple: TupleExprElementListSyntax?) {
+        let elements = pattern.elements.map { $0 }
+        let types: [TupleTypeElementSyntax?] = typeTuple?.map { $0 } ?? Array(repeating: nil, count: elements.count)
+        let initializers: [TupleExprElementSyntax?] = initializerTuple?.map { $0 } ?? Array(repeating: nil, count: elements.count)
+
+        for (element, (type, initializer)) in zip(elements, zip(types, initializers)) {
+            if let elementTuplePattern = element.pattern.as(TuplePatternSyntax.self) {
+                let typeTuple = type?.type.as(TupleTypeSyntax.self)?.elements
+                let initializerTuple = initializer?.expression.as(TupleExprSyntax.self)?.elementList
+
+                visitVariableTupleBinding(
+                    node: node,
+                    pattern: elementTuplePattern,
+                    typeTuple: typeTuple,
+                    initializerTuple: initializerTuple)
+            } else {
+                parse(
+                    modifiers: node.modifiers,
+                    attributes: node.attributes,
+                    trivia: node.leadingTrivia,
+                    variableType: type?.type,
+                    variableInitFunctionCallExpr: initializer?.expression.as(FunctionCallExprSyntax.self),
+                    at: element.positionAfterSkippingLeadingTrivia)
             }
         }
     }
@@ -318,6 +329,7 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
         inheritanceClause: TypeInheritanceClauseSyntax? = nil,
         genericParameterClause: GenericParameterClauseSyntax? = nil,
         genericWhereClause: GenericWhereClauseSyntax? = nil,
+        variableInitFunctionCallExpr: FunctionCallExprSyntax? = nil,
         consumeCapitalSelfFunctionCalls: Bool = false,
         at position: AbsolutePosition
     ) {
@@ -342,6 +354,12 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
             self.didVisitCapitalSelfFunctionCall = false
         }
 
+        let returnClauseTypeLocations = typeNameLocations(for: returnClause)
+        let hasGenericFunctionReturnedMetatypeParameters = hasGenericFunctionReturnedMetatypeParameters(
+            genericParameterClause: genericParameterClause,
+            parameterClause: parameterClause,
+            returnClauseTypeLocations: returnClauseTypeLocations)
+
         results.append((
             location,
             accessibility,
@@ -351,13 +369,45 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
             type(for: variableType),
             typeLocations(for: variableType),
             typeLocations(for: parameterClause),
-            typeLocations(for: returnClause),
+            Set(returnClauseTypeLocations.map { $0.location }),
             typeLocations(for: inheritanceClause),
             typeLocations(for: genericParameterClause),
             typeLocations(for: genericWhereClause),
+            locations(for: variableInitFunctionCallExpr),
+            functionCallMetatypeArgumentLocations(for: variableInitFunctionCallExpr),
             letShorthandIdentifiers,
-            didVisitCapitalSelfFunctionCall
+            didVisitCapitalSelfFunctionCall,
+            hasGenericFunctionReturnedMetatypeParameters
         ))
+    }
+
+    /// Determines whether the function has generic metatype parameters that are also returned.
+    /// For example: `func someFunc<T>(type: T.Type) -> T`.
+    private func hasGenericFunctionReturnedMetatypeParameters(
+        genericParameterClause: GenericParameterClauseSyntax?,
+        parameterClause: ParameterClauseSyntax?,
+        returnClauseTypeLocations: Set<TypeNameSourceLocation>
+    ) -> Bool {
+        guard let genericParameterClause, let parameterClause else { return false }
+
+        let genericParameterNames = genericParameterClause
+            .genericParameterList
+            .reduce(into: Set<String>()) { result, param in
+                result.insert(param.name.trimmedDescription)
+            }
+
+        return parameterClause
+            .parameterList
+            .contains {
+                if let baseTypeName = $0.type?.as(MetatypeTypeSyntax.self)?.baseType.trimmedDescription,
+                   genericParameterNames.contains(baseTypeName),
+                   returnClauseTypeLocations.contains(where: { $0.name == baseTypeName })
+                {
+                    return true
+                }
+
+                return false
+            }
     }
 
     private func type(for typeSyntax: TypeSyntax?) -> String? {
@@ -380,14 +430,14 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
         })
     }
 
-    private func typeLocations(for clause: ReturnClauseSyntax?) -> Set<SourceLocation> {
+    private func typeNameLocations(for clause: ReturnClauseSyntax?) -> Set<TypeNameSourceLocation> {
         guard let returnTypeSyntax = clause?.returnType else { return [] }
 
         if let someReturnType = returnTypeSyntax.as(ConstrainedSugarTypeSyntax.self) {
-            return typeSyntaxInspector.typeLocations(for: someReturnType.baseType)
+            return typeSyntaxInspector.typeNameLocations(for: someReturnType.baseType)
         }
 
-        return typeSyntaxInspector.typeLocations(for: returnTypeSyntax)
+        return typeSyntaxInspector.typeNameLocations(for: returnTypeSyntax)
     }
 
     private func typeLocations(for clause: GenericParameterClauseSyntax?) -> Set<SourceLocation> {
@@ -416,5 +466,27 @@ final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
         return clause.inheritedTypeCollection.reduce(into: .init()) { result, type in
             result.formUnion(typeSyntaxInspector.typeLocations(for: type.typeName))
         }
+    }
+
+    private func locations(for call: FunctionCallExprSyntax?) -> Set<SourceLocation> {
+        guard let call = call else { return [] }
+
+        return [sourceLocationBuilder.location(at: call.positionAfterSkippingLeadingTrivia)]
+    }
+
+    private func functionCallMetatypeArgumentLocations(for call: FunctionCallExprSyntax?) -> Set<SourceLocation> {
+        guard let call else { return [] }
+
+        return call
+            .argumentList
+            .reduce(into: .init(), { result, argument in
+                if let memberExpr = argument.expression.as(MemberAccessExprSyntax.self),
+                   memberExpr.name.tokenKind == .keyword(.`self`),
+                   let baseIdentifier = memberExpr.base?.as(IdentifierExprSyntax.self)
+                {
+                    let location = sourceLocationBuilder.location(at: baseIdentifier.positionAfterSkippingLeadingTrivia)
+                    result.insert(location)
+                }
+            })
     }
 }

--- a/Sources/PeripheryKit/Syntax/TypeSyntaxInspector.swift
+++ b/Sources/PeripheryKit/Syntax/TypeSyntaxInspector.swift
@@ -2,6 +2,11 @@ import Foundation
 import SwiftSyntax
 import Shared
 
+struct TypeNameSourceLocation: Hashable {
+    let name: String
+    let location: SourceLocation
+}
+
 struct TypeSyntaxInspector {
     let sourceLocationBuilder: SourceLocationBuilder
 
@@ -9,55 +14,60 @@ struct TypeSyntaxInspector {
         PropertyTypeSanitizer.sanitize(typeSyntax.description)
     }
 
-    func typeLocations(for typeSyntax: TypeSyntax) -> Set<SourceLocation> {
-        if let simpleType = typeSyntax.as(SimpleTypeIdentifierSyntax.self) {
-            // Simple type.
-            let location = nameTypeLocation(for: simpleType.name)
-            let genericTypeLocations = Set(simpleType.genericArgumentClause?.arguments.flatMap { typeLocations(for: $0.argumentType) } ?? [])
-            return genericTypeLocations.union([location])
-        } else if let optionalType = typeSyntax.as(OptionalTypeSyntax.self) {
-            // Optional type.
-            return typeLocations(for: optionalType.wrappedType)
-        } else if let memberType = typeSyntax.as(MemberTypeIdentifierSyntax.self) {
-            // Member type
-            let location = sourceLocationBuilder.location(at: memberType.name.positionAfterSkippingLeadingTrivia)
-            return typeLocations(for: memberType.baseType).union([location])
-        } else if let tuple = typeSyntax.as(TupleTypeSyntax.self) {
-            // Tuple type.
-            return Set(tuple.elements.flatMap { typeLocations(for: $0.type) })
-        } else if let funcType = typeSyntax.as(FunctionTypeSyntax.self) {
-            // Function type.
-            let locations = funcType.arguments.flatMap { typeLocations(for: $0.type) }
-            return typeLocations(for: funcType.output.returnType).union(locations)
-        } else if let arrayType = typeSyntax.as(ArrayTypeSyntax.self) {
-            // Array type.
-            return typeLocations(for: arrayType.elementType)
-        } else if let dictType = typeSyntax.as(DictionaryTypeSyntax.self) {
-            // Dictionary type.
-            return typeLocations(for: dictType.keyType).union(typeLocations(for: dictType.valueType))
-        } else if let someType = typeSyntax.as(ConstrainedSugarTypeSyntax.self) {
-            // Some type.
-            return typeLocations(for: someType.baseType)
-        } else if let implicitUnwrappedOptionalType = typeSyntax.as(ImplicitlyUnwrappedOptionalTypeSyntax.self) {
-            // Implicitly unwrapped optional type.
-            return typeLocations(for: implicitUnwrappedOptionalType.wrappedType)
-        } else if let compositionType = typeSyntax.as(CompositionTypeSyntax.self) {
-            // Composition type.
-            return Set(compositionType.elements.flatMap { typeLocations(for: $0.type) })
-        } else if let attributedType = typeSyntax.as(AttributedTypeSyntax.self) {
-            // Attributed type.
-            return typeLocations(for: attributedType.baseType)
-        } else if let metatypeType = typeSyntax.as(MetatypeTypeSyntax.self) {
-            // Metatype type.
-            return typeLocations(for: metatypeType.baseType)
+    func typeNameLocations(for typeSyntax: TypeSyntax) -> Set<TypeNameSourceLocation> {
+        types(for: typeSyntax).mapSet {
+            .init(name: $0.trimmedDescription,
+                  location: sourceLocationBuilder.location(at: $0.positionAfterSkippingLeadingTrivia))
         }
+    }
 
-        return [sourceLocationBuilder.location(at: typeSyntax.positionAfterSkippingLeadingTrivia)]
+    func typeLocations(for typeSyntax: TypeSyntax) -> Set<SourceLocation> {
+        types(for: typeSyntax).mapSet { sourceLocationBuilder.location(at: $0.positionAfterSkippingLeadingTrivia) }
     }
 
     // MARK: - Private
 
-    private func nameTypeLocation(for name: TokenSyntax) -> SourceLocation {
-        sourceLocationBuilder.location(at: name.positionAfterSkippingLeadingTrivia)
+    func types(for typeSyntax: TypeSyntax) -> Set<TokenSyntax> {
+        if let simpleType = typeSyntax.as(SimpleTypeIdentifierSyntax.self) {
+            // Simple type.
+            var result = simpleType.genericArgumentClause?.arguments.flatMapSet { types(for: $0.argumentType) } ?? []
+            return result.inserting(simpleType.name)
+        } else if let optionalType = typeSyntax.as(OptionalTypeSyntax.self) {
+            // Optional type.
+            return types(for: optionalType.wrappedType)
+        } else if let memberType = typeSyntax.as(MemberTypeIdentifierSyntax.self) {
+            // Member type.
+            return types(for: memberType.baseType).union([memberType.name])
+        } else if let tuple = typeSyntax.as(TupleTypeSyntax.self) {
+            // Tuple type.
+            return tuple.elements.flatMapSet { types(for: $0.type) }
+        } else if let funcType = typeSyntax.as(FunctionTypeSyntax.self) {
+            // Function type.
+            let argumentTypes = funcType.arguments.flatMapSet { types(for: $0.type) }
+            return types(for: funcType.output.returnType).union(argumentTypes)
+        } else if let arrayType = typeSyntax.as(ArrayTypeSyntax.self) {
+            // Array type.
+            return types(for: arrayType.elementType)
+        } else if let dictType = typeSyntax.as(DictionaryTypeSyntax.self) {
+            // Dictionary type.
+            return types(for: dictType.keyType).union(types(for: dictType.valueType))
+        } else if let someType = typeSyntax.as(ConstrainedSugarTypeSyntax.self) {
+            // Some type.
+            return types(for: someType.baseType)
+        } else if let implicitUnwrappedOptionalType = typeSyntax.as(ImplicitlyUnwrappedOptionalTypeSyntax.self) {
+            // Implicitly unwrapped optional type.
+            return types(for: implicitUnwrappedOptionalType.wrappedType)
+        } else if let compositionType = typeSyntax.as(CompositionTypeSyntax.self) {
+            // Composition type.
+            return Set(compositionType.elements.flatMap { types(for: $0.type) })
+        } else if let attributedType = typeSyntax.as(AttributedTypeSyntax.self) {
+            // Attributed type.
+            return types(for: attributedType.baseType)
+        } else if let metatypeType = typeSyntax.as(MetatypeTypeSyntax.self) {
+            // Metatype type.
+            return types(for: metatypeType.baseType)
+        }
+
+        return []
     }
 }

--- a/Sources/Shared/Extensions/Sequence+Extension.swift
+++ b/Sources/Shared/Extensions/Sequence+Extension.swift
@@ -1,7 +1,7 @@
 import Foundation
 
-extension Sequence {
-    public func mapFirst<T>(_ transform: (Self.Element) throws -> T?) rethrows -> T? {
+public extension Sequence {
+    func mapFirst<T>(_ transform: (Element) throws -> T?) rethrows -> T? {
         for item in self {
             if let transformed = try transform(item) {
                 return transformed
@@ -9,5 +9,17 @@ extension Sequence {
         }
 
         return nil
+    }
+
+    func flatMapSet<T>(_ transform: (Element) throws -> Set<T>) rethrows -> Set<T> {
+        try reduce(into: .init()) { result, element in
+            result.formUnion(try transform(element))
+        }
+    }
+
+    func mapSet<T>(_ transform: (Element) throws -> T) rethrows -> Set<T> {
+        try reduce(into: .init()) { result, element in
+            result.insert(try transform(element))
+        }
     }
 }

--- a/Sources/Shared/Extensions/Set+Extension.swift
+++ b/Sources/Shared/Extensions/Set+Extension.swift
@@ -1,0 +1,8 @@
+import Foundation
+
+public extension Set {
+    mutating func inserting(_ value: Element) -> Self {
+        insert(value)
+        return self
+    }
+}

--- a/Tests/AccessibilityTests/AccessibilityProject/Sources/MainTarget/main.swift
+++ b/Tests/AccessibilityTests/AccessibilityProject/Sources/MainTarget/main.swift
@@ -6,9 +6,7 @@ PublicExtensionOnRedundantPublicKindRetainer().retain()
 IgnoreCommentCommandRetainer().retain()
 IgnoreAllCommentCommandRetainer().retain()
 
-_ = PublicTypeUsedAsPublicPropertyTypeRetainer().retain1
-_ = PublicTypeUsedAsPublicPropertyTypeRetainer().retain2
-_ = PublicTypeUsedAsPublicPropertyTypeRetainer().retain3
+_ = PublicTypeUsedAsPublicPropertyTypeRetainer().retain
 
 _ = PublicTypeUsedAsPublicInitializerParameterTypeRetainer()
 
@@ -28,6 +26,8 @@ PublicTypeUsedAsPublicFunctionGenericParameterRetainer().retain(PublicTypeUsedAs
 PublicTypeUsedAsPublicFunctionGenericRequirementRetainer().retain(PublicTypeUsedAsPublicFunctionGenericRequirement_ConformingClass.self)
 _ = PublicTypeUsedAsPublicClassGenericParameterRetainer<PublicTypeUsedAsPublicClassGenericParameter_ConformingClass>()
 _ = PublicTypeUsedAsPublicClassGenericRequirementRetainer<PublicTypeUsedAsPublicClassGenericRequirement_ConformingClass>()
+
+_ = PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnTypeRetainer().retain
 
 _ = NotRedundantPublicTestableImportClass().testableProperty
 

--- a/Tests/AccessibilityTests/AccessibilityProject/Sources/TargetA/PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType.swift
+++ b/Tests/AccessibilityTests/AccessibilityProject/Sources/TargetA/PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType.swift
@@ -1,0 +1,59 @@
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType1 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType2 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType3 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType4 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType5 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType6 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType7 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType8 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType9 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_1 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_2 {}
+public protocol PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_3 {}
+
+public class PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnTypeRetainer {
+    public init() {}
+
+    public var retain: [Any] {
+        [retain1, retain2, retain3, retain4, retain5, retain6, retain7, retain8, retain9, retain10]
+    }
+
+    public let retain1 = simpleReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType1.self)
+    public let (retain2, retain3, retain4) = (
+        simpleReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType2.self),
+        simpleReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType3.self),
+        nonMetatypeReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType4.self)
+    )
+    public let (retain5, (retain6, retain7)) = (simpleReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType5.self), (simpleReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType6.self), simpleReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType7.self)))
+    public let retain8 = closureReturn(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType8.self)
+    public let retain9 = closureParameter(PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType9.self)
+    public let retain10 = multipleParameters(
+        PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_1.self,
+        PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_2.self,
+        PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_3.self
+    )
+
+    private static func closureReturn<T>(_ type: T.Type) -> () -> T {
+        fatalError()
+    }
+
+    private static func closureParameter<T>(_ type: T.Type) -> (T) -> Void {
+        fatalError()
+    }
+
+    private static func simpleReturn<T>(_ type: T.Type) -> T {
+        fatalError()
+    }
+
+    /// Even though C is not returned, it's assumed to be publicly accessible because it's too
+    /// complex to solve for the scenario where the call site argument positions do not match the
+    /// function signature parameter positions. Parameters with default arguments can result in
+    /// misalignment.
+    private static func multipleParameters<A, B, C>(_ typeA: A.Type, _ typeB: B.Type, _ typeC: C.Type) -> (A, B) {
+        fatalError()
+    }
+
+    private static func nonMetatypeReturn<T>(_ type: T.Type) -> String {
+        ""
+    }
+}

--- a/Tests/AccessibilityTests/AccessibilityProject/Sources/TargetA/PublicTypeUsedAsPublicPropertyType.swift
+++ b/Tests/AccessibilityTests/AccessibilityProject/Sources/TargetA/PublicTypeUsedAsPublicPropertyType.swift
@@ -1,12 +1,21 @@
 import Foundation
 
-public class PublicTypeUsedAsPublicPropertyType {}
+public class PublicTypeUsedAsPublicPropertyType1 {}
+public class PublicTypeUsedAsPublicPropertyType2 {}
+public class PublicTypeUsedAsPublicPropertyType3 {}
+public class PublicTypeUsedAsPublicPropertyType4 {}
 public struct PublicTypeUsedAsPublicPropertyGenericArgumentType: Hashable {}
 public class PublicTypeUsedAsPublicPropertyArrayType {}
 
 public class PublicTypeUsedAsPublicPropertyTypeRetainer {
     public init() {}
-    public var retain1: PublicTypeUsedAsPublicPropertyType?
+
+    public var retain: [Any?] {
+        [retain1, retain2, retain3, retain4, retain5, retain6]
+    }
+
+    public var retain1: PublicTypeUsedAsPublicPropertyType1?
     public var retain2: Set<PublicTypeUsedAsPublicPropertyGenericArgumentType>?
     public var retain3: [PublicTypeUsedAsPublicPropertyArrayType] = []
+    public var (retain4, (retain5, retain6)): (PublicTypeUsedAsPublicPropertyType2, (PublicTypeUsedAsPublicPropertyType3, PublicTypeUsedAsPublicPropertyType4)) = (.init(), (.init(), .init()))
 }

--- a/Tests/AccessibilityTests/RedundantPublicAccessibilityTest.swift
+++ b/Tests/AccessibilityTests/RedundantPublicAccessibilityTest.swift
@@ -31,7 +31,10 @@ class RedundantPublicAccessibilityTest: SourceGraphTestCase {
     }
 
     func testPublicTypeUsedAsPublicPropertyType() {
-        assertNotRedundantPublicAccessibility(.class("PublicTypeUsedAsPublicPropertyType"))
+        assertNotRedundantPublicAccessibility(.class("PublicTypeUsedAsPublicPropertyType1"))
+        assertNotRedundantPublicAccessibility(.class("PublicTypeUsedAsPublicPropertyType2"))
+        assertNotRedundantPublicAccessibility(.class("PublicTypeUsedAsPublicPropertyType3"))
+        assertNotRedundantPublicAccessibility(.class("PublicTypeUsedAsPublicPropertyType4"))
         assertNotRedundantPublicAccessibility(.struct("PublicTypeUsedAsPublicPropertyGenericArgumentType"))
         assertNotRedundantPublicAccessibility(.class("PublicTypeUsedAsPublicPropertyArrayType"))
     }
@@ -143,5 +146,22 @@ class RedundantPublicAccessibilityTest: SourceGraphTestCase {
     func testPublicTypeUsedInPublicClosure() {
         assertNotRedundantPublicAccessibility(.class("PublicTypeUsedInPublicClosureReturnType"))
         assertNotRedundantPublicAccessibility(.class("PublicTypeUsedInPublicClosureInputType"))
+    }
+
+    func testFunctionMetatypeParameterUsedAsGenericReturnType() {
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType1"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType2"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType3"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType5"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType6"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType7"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType8"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType9"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_1"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_2"))
+        assertNotRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType10_3"))
+
+        // Destructured binding control.
+        assertRedundantPublicAccessibility(.protocol("PublicTypeUsedAsPublicFunctionMetatypeParameterWithGenericReturnType4"))
     }
 }

